src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/MetadataCollection.cs

```diff
@@ -160,18 +153,17 @@ public static MetadataCollection Create(Dictionary<string, string?> map)
             return new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value, pair3.Key, pair3.Value);
         }
 
-        // Finally, if there are four or more items, add the pairs to a list in order to construct
+        // Finally, if there are four or more items, add the pairs to an array in order to construct
         // a FourOrMoreItems instance. Note that the constructor will copy the key-value pairs and won't
-        // hold onto the list we're passing, so it's safe to use a pooled list.
-        using var _ = ListPool<KeyValuePair<string, string?>>.GetPooledObject(out var list);
-        list.SetCapacityIfLarger(count);
+        // hold onto the list we're passing, so it's safe to use a pooled array.
+        using var array = new PooledSpanBuilder<KeyValuePair<string, string?>>(count);
```

Comment:
Could use `MemoryBuilder<KeyValuePair<string, string?>>`.

Alternatively, you could use a `PooledArrayBuilder`, add a new `Create(ref readonly PooledArrayBuilder<...>)` and just pass the builder in. That way, small arrays are stored on the stack and don't impact pools at all.